src/vmm/src/builder.rs

```diff
@@ -711,7 +711,8 @@ pub fn setup_serial_device(
         ),
         input: Some(input),
     })));
-    event_manager.add_subscriber(serial.clone());
+    event_manager.add_subscriber(Arc::new(Mutex::new(serial.clone()))); // TODO: Check this actually
```

Comment:
should just be 

```suggestion
    event_manager.add_subscriber(Arc::clone(&serial)); 
```

I think

Reply:
Is there any semantic difference between `Arc::clone(&serial)` and `serial.clone()`? The latter seems to be accepted here, but the former is not.

Reply:
Mhhh, wow, okay, my turn to be absolutely stumped by what the rust compiler wants. The following works:

```diff
diff --git a/src/vmm/src/builder.rs b/src/vmm/src/builder.rs
index 27d986251..42055b592 100644
--- a/src/vmm/src/builder.rs
+++ b/src/vmm/src/builder.rs
@@ -723,7 +723,8 @@ pub fn setup_serial_device(
     let interrupt_evt = EventFdTrigger::new(EventFd::new(EFD_NONBLOCK).map_err(VmmError::EventFd)?);
     let kick_stdin_read_evt =
         EventFdTrigger::new(EventFd::new(EFD_NONBLOCK).map_err(VmmError::EventFd)?);
-    let serial = BusDevice::Serial(Arc::new(Mutex::new(SerialWrapper {
+
+    let serial = Arc::new(Mutex::new(SerialWrapper {
         serial: Serial::with_events(
             interrupt_evt,
             SerialEventsWrapper {
@@ -732,10 +733,13 @@ pub fn setup_serial_device(
             SerialOut::Stdout(out),
         ),
         input: Some(input),
-    })));
-    event_manager.add_subscriber(Arc::new(Mutex::new(serial.clone()))); // TODO: Check this actually
-                                                                        // works?
-    Ok(serial)
+    }));
+
+    let to_return = Arc::clone(&serial);
+
+    event_manager.add_subscriber(serial);
+
+    Ok(BusDevice::Serial(to_return))
 }
```

but changing the last 3 lines to be 

```
event_manager.add_subscriber(Arc::clone(&serial));
Ok(BusDevice::Serial(to_return))
```

doesn't compile, with it complaining that `serial` doesnt implement `MutEventSubscriber` (even though it does). 

Anyway, in my original comment, I missed that `serial` is a `BusDevice` instead of a `SerialWrapper`, so your solution is okay (as you're still just cloning an Arc instead of the actual device). However, it does end up registering something like `Arc<Mutex<Arc<Mutex<SerialWrapper>>>>` to the event bus.

Reply:
Ah, and you've pretty much already done exactly what I suggested, lol. Should've checked the diff before commenting, sorry haha. Just having the `.clone()` is fine, although for these smart pointer types that `Deref` into their pointed-to value, writing `Arc::clone()` clarifies that we're cloning the `Arc` and not what's inside the `Arc`. Definitely more of a nit though, no need to fight rustc over this :)